macaroonbakery/discharge.py

```diff
@@ -0,0 +1,193 @@
+    :param locator: used to information on third parties referred to by third
+    party caveats returned by the Checker.
+    '''
+    caveat_id_prefix = []
+    if caveat is None:
+        # The caveat information is encoded in the id itself.
+        caveat = id
+    else:
+        # We've been given an explicit id, so when extra third party
+        # caveats are added, use that id as the prefix
+        # for any more ids.
+        caveat_id_prefix = id
+    cav_info = decode_caveat(key, caveat)
+    cav_info.id = id
+    # Note that we don't check the error - we allow the
+    # third party checker to see even caveats that we can't
+    # understand.
+    cond, arg = parse_caveat(cav_info.condition)
+
+    if cond == COND_NEED_DECLARED:
+        cav_info.condition = arg.encode('utf-8')
+        caveats = _check_need_declared(ctx, cav_info, checker)
+    else:
+        caveats = checker.check_third_party_caveat(ctx, cav_info)
+
+    # Note that the discharge macaroon does not need to
+    # be stored persistently. Indeed, it would be a problem if
+    # we did, because then the macaroon could potentially be used
+    # for normal authorization with the third party.
+    m = Macaroon(cav_info.root_key, id, '', cav_info.version,
+                 cav_info.ns)
+    m.caveat_id_prefix = caveat_id_prefix
+    if caveats is not None:
+        for cav in caveats:
+            m.add_caveat(ctx, cav, key, locator)
+    return m
+
+
+def _check_need_declared(ctx, cav_info, checker):
+    arg = cav_info.condition.decode('utf-8')
+    i = arg.find(' ')
+    if i <= 0:
+        raise ValueError(
+            'need-declared caveat requires an argument, got %q'.format(arg))
+    need_declared = arg[0:i].split(',')
+    for d in need_declared:
+        if d == '':
+            raise ValueError('need-declared caveat with empty required '
+                             'attribute')
+    if len(need_declared) == 0:
+        raise ValueError('need-declared caveat with no required attributes')
+    cav_info.condition = arg[i+1:].encode('utf-8')
+    caveats, err = checker.check_third_party_caveat(ctx, cav_info)
+    declared = {}
+    for cav in caveats:
+        if cav.location != '':
+            continue
+        # Note that we ignore the error. We allow the service to
+        # generate caveats that we don't understand here.
+        try:
+            cond, arg = parse_caveat(cav.condition)
+        except ValueError:
+            continue
+        if cond != COND_DECLARED:
+            continue
+        parts = arg.split()
+        if len(parts) != 2:
+            raise ValueError('declared caveat has no value')
+        declared[parts[0]] = True
+    # Add empty declarations for everything mentioned in need-declared
+    # that was not actually declared.
+    for d in need_declared:
+        if not declared[d]:
+            caveats.append(declared_caveat(d, ''))
+    return caveats
+
+
+class _EmptyLocator(object):
```

Comment:
shouldn't this inherit from the ThirdPartyLocator abstract base class?

Reply:
done.